### PR TITLE
optimize EnvoyFilter rebuild

### PR DESCRIPTION
**Pull request by sschepens**
**Please provide a description of this PR:**
We're seeing a big time spent in rebuilding `EnvoyFilter` indexes, this is because converting a Struct to it's corresponding Proto requires JSON marshalling and unmarshalling and is very expensive.

I don't think there's a better solution to converting a Struct to it's corresponding Proto, but what I do think is worth doing is avoid converting all of them everytime any single `EnvoyFilter` gets updated.

The idea is to create a shallow copy of the previous `EnvoyFilter` index and only rebuild EnvoyFilters which have actually changed, this should singificantly reduce the amount of work done .

Proof:
```
(pprof) top10 -cum
Showing nodes accounting for 0.01s, 0.036% of 27.41s total
Dropped 588 nodes (cum <= 0.14s)
Showing top 10 nodes out of 227
      flat  flat%   sum%        cum   cum%
         0     0%     0%     18.49s 67.46%  istio.io/istio/pilot/pkg/model.(*PushContext).updateContext
         0     0%     0%     18.40s 67.13%  istio.io/istio/pilot/pkg/model.(*PushContext).InitContext
         0     0%     0%     18.32s 66.84%  istio.io/istio/pilot/pkg/xds.(*DiscoveryServer).initPushContext
         0     0%     0%     18.26s 66.62%  istio.io/istio/pilot/pkg/xds.(*DiscoveryServer).Push
         0     0%     0%     18.23s 66.51%  istio.io/istio/pilot/pkg/xds.debounce.func1
         0     0%     0%     14.37s 52.43%  istio.io/istio/pkg/config/xds.StructToMessage
         0     0%     0%     14.31s 52.21%  istio.io/istio/pkg/config/xds.BuildXDSObjectFromStruct
     0.01s 0.036% 0.036%     14.23s 51.92%  istio.io/istio/pilot/pkg/model.convertToEnvoyFilterWrapper
         0     0% 0.036%     14.21s 51.84%  istio.io/istio/pilot/pkg/model.(*PushContext).initEnvoyFilters
         0     0% 0.036%     12.66s 46.19%  github.com/golang/protobuf/jsonpb.(*Unmarshaler).Unmarshal (inline)
```

Just missing some tests but want to get a feeling if this is OK.

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
We have already optimized push context update, only update envoyfilters when envoyfilter updates. Haven't look how `a shared concurrent map ` works, and how much memory increased

**Comment by sschepens** [comment]
> We have already optimized push context update, only update envoyfilters when envoyfilter updates. Haven't look how `a shared concurrent map ` works, and how much memory increased

Of course envoy filters are only built when something has changed, but it **rebuilds all** envoy filters even if **only one** has changed, this behaviour is less than ideal when having many envoy filters.

A shared concurrent map will have a bit of memory impact but probably not too much since we were already storing this information in a map and were creating new maps everytime.
On the other hand a shard concurrent map will be much better in terms of contention than using a regular map with a mutex.

**Comment by sschepens** [comment]
are istio integration tests runnable locally? i´m having trouble having them run in a local k8s cluster on mac because it attempts to start both an `ingressgateway` and a `eastwestgateway` with the same nodeport (15021)

**Comment by hzxuzhonghu** [comment]
How many envoyfilters do you have? I am wondering how much it influenced us, we do not want to over-optimize without seeing much improvement

**Comment by sschepens** [comment]
We have 2k EnvoyFilters and it will continue to grow.

I really don't think this is over-optimizing, it's actually what should probably be done for every resource, it's a waste of compute to be rebuilding all of the indexes everytime a single instane of a resource has changed. I intend to eventually extend this mechanism to other resources.

**Review by howardjohn — CHANGES_REQUESTED** [review]
I don't think this is safe. PushContext is a snapshot. If we mutate the snapshot, we are mutating across all versions - including old ones which should not be changed.

**Comment by sschepens** [comment]
@howardjohn I know, but is there really a problem with mutating the index across all versions? I mean, it's probale that an earlier versions sees a new change and is impacted by it, but could this cause issues? If there were issues, a new version is actually being created by this new PushContext and so everything should eventually end up OK. This would make the system more eventually-consistent but I don't know if it could really cause issues.

Otherwise, is there a way we can get around this? I can imagine a few options:
- maintain a shared index in another place and create a copy of the index for every push context: this could still prove expensive for large deployments since we would need to copy big maps
- add some sort of mechanisms (locks?) to prevent updates when older push contexts are reading the index and prevent them from further reading the index until the new pushcontext is created. Is there any reason why older PushContext should be able to continue to operate if we're actually creating a new PushContext?

**Comment by howardjohn** [comment]
Yes, the entire XDS generation is reliant on PushContext being a snapshot.
There are tests that will catch it not being a snapshot as well, it's just
pretty tricky to exercise so likely will not show up with a simple test run.

I don't think copying would be too expensive, isn't it just a map of 2000
pointers basically?

> Is there any reason why older PushContext should be able to continue to
operate if we're actually creating a new PushContext?

Yes, ongoing pushes from the old context

**Comment by sschepens** [comment]
> I don't think copying would be too expensive, isn't it just a map of 2000
pointers basically?

I'll modify the PR to do this, but we will probably have many more in the future, we can attempt to further optimize then

> operate if we're actually creating a new PushContext?
Yes, ongoing pushes from the old context

What I meant was, couln't we hold off creating a new PushContext until pushes from the old context are finished and then "block" that old push context until we create the new one that's going to replace it? 🤔

**Comment by howardjohn** [comment]
I will take a look at the new changes soon, in the meantime some similar discussions: 

https://github.com/istio/istio/pull/45733#discussion_r1254671346

https://docs.google.com/document/d/1-ywpCnOfubqg7WAXSPf4YgbaFDBEU9HIqMWcxZLhzwE/edit

**Comment by sschepens** [comment]
@jacob-delgado I know, and we certainly don't want 2k EnvoyFilters, but there are a lot of things not covered by Istio networking APIs. Some things we do:
- EnvoyFilter configuring RBAC filter because Istio doesn't want to add regex capabilites and we use them. There is an open issue about this that has a long time.
- EnvoyFilter to create an Ingress route. (there's currently no way of defining ingress routes, this is something sort of legacy and should not be mantained in the future, so i'm not too worried about this).
- EnvoyFilter to add some headers to all outgoing or incoming requests. (there's also no support for this type of functionality, some are custom header sanitizations, others are informative headers, others are, for example authorization headers for legacy non-mesh workloads).

Regardless of the purposes, I think the optimizations should be welcome and should probably be ported to the handling of every resource type.

**Comment by sschepens** [comment]
also, RBAC configurations tend to be very large and verbose, so rebuilding all of these all the time is very costly

File: pilot/pkg/model/push_context.go
Comment (hzxuzhonghu) [inline pilot/pkg/model/push_context.go:1915]: In this way, we will need to handle add/update and delete. This is a little too complex, I am concerned it will not improve when the number of changed envoyfiter is large

Comment (sschepens) [reply pilot/pkg/model/push_context.go:1915]: for people with little amount of EnvoyFilters, which is probably much more common, the difference should be neglegible.
for poeple with a huge amount of EnvoyFilters, it's probably impossible to update a huge amount of them in a single debounce window.

I really don't see any realistic scenario where this changes could end up degrading performance instead of improving it, but i can feature flag this behaviour if that sounds better.

**Comment by howardjohn** [comment]
> Regardless of the purposes, I think the optimizations should be welcome and should probably be ported to the handling of every resource type.

FWIW I understand the intent of this statement but I would necessarily agree in general. We are seeing a similar issue with CDS Delta implementation - the complexity of each additional one is becoming too much to reasonably maintain without doing a more holistic re-architecture

**Comment by sschepens** [comment]
> FWIW I understand the intent of this statement but I would necessarily agree in general. We are seeing a similar issue with CDS Delta implementation - the complexity of each additional one is becoming too much to reasonably maintain without doing a more holistic re-architecture

I agree that a more holistic re-architecture would be the best way to solve these things and I really like the document you shared, but I think those approaches are probably far away enough to justify optimizing these flows in the mean time.

**Comment by howardjohn** [comment]
I totally agree - just would be concerned about merging 10-20 PRs of this
scope. Hard to find the right balance, though

File: pilot/pkg/model/fake_store.go
Comment (howardjohn) [inline pilot/pkg/model/fake_store.go:75]: nit:

`return maps.Values(nsConfigs[namespace])`

File: pilot/pkg/model/push_context.go
Comment (howardjohn) [inline pilot/pkg/model/push_context.go:1919]:
```suggestion
		ps.envoyFiltersByNamespace[k] = slices.Clone(v)
```

Seems preferable to do it here instead of L1928; the latter is in a loop?

I suppose you avoid copy for namespaces we don't modify, although its just a copy of the pointers so not terrible. We could precompute the set of namespaces changed and only clone them if we really want.

if not, these 4 lines can just be `ps.envoyFiltersByNamespace = maps.Clone(previousIndex)` btw

File: pilot/pkg/model/push_context.go
Comment (howardjohn) [reply pilot/pkg/model/push_context.go:1915]: One approach I was thinking of was to combine this with `initEnvoyFilters`. allow it to take in `prev` and `updated`.

Pseudo code:

```go
now := env.List(EnvoyFilter)
res := make([]config, 0, len(now))
for _, cfg := range now {
	if changed.Contains(cfg) {
		// It changed, recompute
		res = append(res, computeEnvoyFilter(cfg))
	} else if prev.Contains(cfg) {
		// Not changed and we already have it, grab the previous one
		res = append(res, prev.Get(cfg))
	} else {
		// New config, compute it
		res = append(res, computeEnvoyFilter(cfg))
	}
}
```

When we currently call initEnvoyFilters we just pass nil for each of the others. This would involve converting `changed` and `prev` to a map of some sort so we can do a lookup, but seems a bit simpler?

Comment (sschepens) [reply pilot/pkg/model/push_context.go:1915]: I was thinking that probably Listing all envoy filters all the time is probably much more expensive than just specifically getting the required envoyfilters, `List` allocates a big slice to hold all envoyfilters. Plus the conversions of `changed` and `prev`.

I could definitely take that approach, just wondering if it's not better to keep it a bit more complicated, but more efficient. I could probably separate the current logic in some methods to keep it simpler.

Comment (howardjohn) [reply pilot/pkg/model/push_context.go:1915]: I have some doubts the `List` call is so bad. Is it a guess or do we observe that?

The reason I ask is we do a List() call on basically everything, so I would expect it to blow up much sooner. Allocating a list of 2k elements, which is basically what List does, is pretty cheap I would think

File: pilot/pkg/model/push_context.go
Comment (sschepens) [reply pilot/pkg/model/push_context.go:1919]: yes, i was intentionally doing it in the loop to avoid copying slices we wont alter

**Comment by ramaraochavali** [comment]
If we decide to merge this, please add a feature flag for this like ENABLE_OPTIMIZED_CONFIG_REBUILD or some thing so that we have a fallback - as this is little complicated and might have some bugs. We can use the same flag for other configs if needed

File: pilot/pkg/model/push_context.go
Comment (sschepens) [reply pilot/pkg/model/push_context.go:1915]: It's really a guess, i know it's cheap to allocate that amount of elements, but it will get bigger eventually, and it's probably putting some unnecessary stress on the GC.
I wouldn't take 2k elements as a strict number, we will probably have much more, unless we can incorporate some of our needs in Istio APIs.

**Comment by sschepens** [comment]
@ramaraochavali added a flag for the new behaviour

@howardjohn i pushed some changes, using more of the `slices` package, hopefully it's easier to read

File: pilot/pkg/model/push_context.go
Comment (ramaraochavali) [inline pilot/pkg/model/push_context.go:1921]: Can you please add tests for updateEnvoyFilters covering various scenarios?

File: pilot/pkg/model/push_context.go
Comment (howardjohn) [reply pilot/pkg/model/push_context.go:1921]: I did a quick benchmark, 0.2ms to allocate 200k size slice. if you get anywhere near that limit things are going to blow up for other reasons anyways. can we please switch to this approach? Its just as efficient, consistent with the rest of the code, and minimizes risk and complexity.

File: pilot/pkg/model/push_context.go
Comment (sschepens) [reply pilot/pkg/model/push_context.go:1958]: i cannt reuse the same function because they take different objects, I could parametrize all the needed fields, but that would mean a method taking like 8 or 10 parameters, didn't seem too good.

File: pilot/pkg/model/push_context.go
Comment (sschepens) [reply pilot/pkg/model/push_context.go:1921]: will do

File: pilot/pkg/model/push_context.go
Comment (howardjohn) [reply pilot/pkg/model/push_context.go:1921]: Its one list call for all namespaces in my recommendation - as we do today - not per namespace

Comment (sschepens) [reply pilot/pkg/model/push_context.go]: yes, but the approach you suggest still involves rebuilding the whole index, we would really only avoid computing the envoyfilter, it's definitely better than what we currently do, but I don't think it's as efficient as copying the entire previous index and only doing the necessary changes to it.

File: pilot/pkg/model/push_context.go
Comment (howardjohn) [inline pilot/pkg/model/push_context.go:1312]: Does this work? won't it insert into a nil map?

File: pilot/pkg/features/pilot.go
Comment (howardjohn) [inline pilot/pkg/features/pilot.go:667]: we can default to `true`, just good to have an opt-out for possibly risky features in case they cause issues

**Review by howardjohn — APPROVED** [review]

**Comment by howardjohn** [comment]
LGTM! Thanks for the followsup, this looks like a great improvement. Just 2 small comments but should be good to go

**Comment by hzxuzhonghu** [comment]
agree, we have no tests for initEnvoyFilter

File: pilot/pkg/model/push_context.go
Comment (sschepens) [reply pilot/pkg/model/push_context.go:1312]: hmm probably doesn't, will initialize set accordingly

**Comment by sschepens** [comment]
@ramaraochavali @hzxuzhonghu added some tests, please see if they are OK

**Review by howardjohn — APPROVED** [review]

**Comment by howardjohn** [comment]
@ramaraochavali @hzxuzhonghu please remove the hold if you are ok with this

File: pilot/pkg/model/push_context.go
Comment (hzxuzhonghu) [inline pilot/pkg/model/push_context.go:1975]: comment this envoyfilter  is  already exist and not changed

File: pilot/pkg/model/push_context.go
Comment (hzxuzhonghu) [inline pilot/pkg/model/push_context.go:1978]: comment for this case it is new added one

File: pilot/pkg/features/pilot.go
Comment (hzxuzhonghu) [inline pilot/pkg/features/pilot.go:667]: Do we want to reuse this flag for other opetimizations later

Comment (ramaraochavali) [reply pilot/pkg/features/pilot.go:667]: Yes. That is the idea if we implement similar optimization for other configs

File: pilot/pkg/model/push_context.go
Comment (ramaraochavali) [reply pilot/pkg/model/push_context.go:1978]: Isn't newly added one covered by L1972(part of changed)? Not sure when this condition will be hit

File: releasenotes/notes/update-envoy-filters-in-place.yaml
Comment (ramaraochavali) [inline releasenotes/notes/update-envoy-filters-in-place.yaml:7]: nit: insted -> instead

File: pilot/pkg/model/push_context_test.go
Comment (ramaraochavali) [inline pilot/pkg/model/push_context_test.go:614]: I think the test case would be lot easier to understand and validate if you have expectedEnvoyFilters in the new push context

File: pilot/pkg/model/push_context.go
Comment (hzxuzhonghu) [reply pilot/pkg/model/push_context.go:1978]: yep, you are right.

This code can be run when `OptimizedConfigRebuild = false`

Comment (sschepens) [reply pilot/pkg/model/push_context.go:1978]: it could happen if `OptimizedConfigRebuild == false` or if for some reason we get an element that wasn't previously there, this should probably never happen, but best to have it in place.

File: pilot/pkg/model/push_context_test.go
Comment (sschepens) [reply pilot/pkg/model/push_context_test.go:614]: i thought aobut it, but I think't it would be very complex as well, because need to check for updates that change the model. `expectedEnvoyFilters` should have defined the entire model to match and would be very extensive, wouldn't it?

File: releasenotes/notes/update-envoy-filters-in-place.yaml
Comment (sschepens) [reply releasenotes/notes/update-envoy-filters-in-place.yaml]: woops, fixed

File: pilot/pkg/model/push_context.go
Comment (sschepens) [reply pilot/pkg/model/push_context.go]: done

File: pilot/pkg/model/push_context.go
Comment (ramaraochavali) [reply pilot/pkg/model/push_context.go:1979]: I think it should never happen. But sure keeping it is safe. Can you add comments to that affect?

File: pilot/pkg/model/push_context_test.go
Comment (ramaraochavali) [reply pilot/pkg/model/push_context_test.go:614]: Is n't it roughly equivalent to initialEnvoyFilters +/- some few (the deleted ones)?

Comment (sschepens) [reply pilot/pkg/model/push_context_test.go:614]: it also does not allow us to test that envoyfilters are exactly the same and have not been rebuilt 🤔

File: pilot/pkg/model/push_context.go
Comment (sschepens) [reply pilot/pkg/model/push_context.go:1979]: done

File: pilot/pkg/model/push_context_test.go
Comment (ramaraochavali) [reply pilot/pkg/model/push_context_test.go:614]: If we provide `expectedEnvoyFiltersByNamespace` for each test case, we can compare with pc2.EnvoyFiltersByNamespace (exactly same as what you are doing). So as long as they are same - our update logic should have worked. The reason why I am thinking about this is we can run the same test case with enable/disable so that the end result is we get the same set of envoy filters.

**Review by ramaraochavali — APPROVED** [review]
Thanks for patiently working through this. This is a good improvement. LGTM, except for one minor test suggestion. But not going to block this PR for that. 
@hzxuzhonghu Please approve and remove hold if it looks good to you.

**Review by hzxuzhonghu — APPROVED** [review]

**Pull request merged**
